File: Liftr/StrengthWorkoutSaveRPC.swift
import Foundation
import Supabase

struct StrengthWorkoutSaveResult: Decodable {
    let workout_id: Int
    let user_id: UUID
    let kind: String
    let title: String?
    let started_at: Date?
    let ended_at: Date?
    let state: String?
    let score: Double?
}

struct StrengthWorkoutFinishLinkedInput: Encodable {
    let workout_id: Int
    let exercises: [StrengthWorkoutExerciseSaveInput]
}

struct StrengthWorkoutExerciseSaveInput: Encodable {
    let workout_exercise_id: Int
    let exercise_id: Int?
    let order_index: Int?
    let notes: String?
    let custom_name: String?
    let sets: [StrengthWorkoutSetSaveInput]
}

struct StrengthWorkoutSetSaveInput: Encodable {
    let set_number: Int
    let order_index: Int?
    let reps: Int?
    let weight_kg: Double?
    let rpe: Double?
    let rest_sec: Int?
    let weight_segments: [StrengthWeightSegWire]?
}

private struct UpdateStrengthWorkoutV1Params: Encodable {
    let p_workout_id: Int
    let p_title: String?
    let p_notes: String?
    let p_started_at: String
    let p_ended_at: String?
    let p_perceived_intensity: String?
    let p_exercises: [StrengthWorkoutExerciseSaveInput]

    private enum CodingKeys: String, CodingKey {
        case p_workout_id
        case p_title
        case p_notes
        case p_started_at
        case p_ended_at
        case p_perceived_intensity
        case p_exercises
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(p_workout_id, forKey: .p_workout_id)
        try container.encode(p_started_at, forKey: .p_started_at)
        try container.encode(p_exercises, forKey: .p_exercises)
        if let p_title {
            try container.encode(p_title, forKey: .p_title)
        } else {
            try container.encodeNil(forKey: .p_title)
        }
        if let p_notes {
            try container.encode(p_notes, forKey: .p_notes)
        } else {
            try container.encodeNil(forKey: .p_notes)
        }
        if let p_ended_at {
            try container.encode(p_ended_at, forKey: .p_ended_at)
        } else {
            try container.encodeNil(forKey: .p_ended_at)
        }
        if let p_perceived_intensity {
            try container.encode(p_perceived_intensity, forKey: .p_perceived_intensity)
        } else {
            try container.encodeNil(forKey: .p_perceived_intensity)
        }
    }
}

private struct FinishStrengthWorkoutV1Params: Encodable {
    let p_workout_id: Int
    let p_ended_at: String
    let p_paused_sec: Int
    let p_exercises: [StrengthWorkoutExerciseSaveInput]
    let p_linked: [StrengthWorkoutFinishLinkedInput]
}

enum StrengthWorkoutSaveRPC {
    static func updateStrengthWorkoutV1(
        client: SupabaseClient,
        workoutId: Int,
        title: String?,
        notes: String?,
        startedAt: String,
        endedAt: String?,
        perceivedIntensity: String?,
        exercises: [StrengthWorkoutExerciseSaveInput]
    ) async throws -> StrengthWorkoutSaveResult {
        let params = UpdateStrengthWorkoutV1Params(
            p_workout_id: workoutId,
            p_title: title,
            p_notes: notes,
            p_started_at: startedAt,
            p_ended_at: endedAt,
            p_perceived_intensity: perceivedIntensity,
            p_exercises: exercises
        )
        let res = try await client
            .rpc("update_strength_workout_v1", params: params)
            .execute()
        return try JSONDecoder.supabase().decode(StrengthWorkoutSaveResult.self, from: res.data)
    }

    static func finishStrengthWorkoutV1(
        client: SupabaseClient,
        workoutId: Int,
        endedAt: String,
        pausedSec: Int,
        exercises: [StrengthWorkoutExerciseSaveInput],
        linked: [StrengthWorkoutFinishLinkedInput]
    ) async throws -> [StrengthWorkoutSaveResult] {
        let params = FinishStrengthWorkoutV1Params(
            p_workout_id: workoutId,
            p_ended_at: endedAt,
            p_paused_sec: pausedSec,
            p_exercises: exercises,
            p_linked: linked
        )
        let res = try await client
            .rpc("finish_strength_workout_v1", params: params)
            .execute()
        return try JSONDecoder.supabase().decode([StrengthWorkoutSaveResult].self, from: res.data)
    }
}

File: android/app/src/main/java/com/lilru/liftr/workout/StrengthWorkoutSaveRpc.kt
package com.lilru.liftr.workout

import com.lilru.liftr.data.BackendContracts
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

@Serializable
data class StrengthWorkoutSaveResultWire(
    @SerialName("workout_id") val workoutId: Int,
    @SerialName("user_id") val userId: String? = null,
    val kind: String? = null,
    val title: String? = null,
    @SerialName("started_at") val startedAt: String? = null,
    @SerialName("ended_at") val endedAt: String? = null,
    val state: String? = null,
    val score: Double? = null
)

internal object StrengthWorkoutSaveRpc {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun updateStrengthWorkoutV1(
        supabase: SupabaseClient,
        workoutId: Int,
        title: String?,
        notes: String?,
        startedAtIso: String,
        endedAtIso: String?,
        perceivedIntensity: String?,
        exercises: List<StrengthEditExercisePayload>
    ): StrengthWorkoutSaveResultWire {
        val params = buildJsonObject {
            put("p_workout_id", workoutId)
            put("p_started_at", startedAtIso)
            put("p_exercises", StrengthWorkoutEditSavePayload.exercisesToJsonArray(exercises))
            if (title != null) put("p_title", title) else put("p_title", JsonNull)
            if (notes != null) put("p_notes", notes) else put("p_notes", JsonNull)
            if (endedAtIso != null) put("p_ended_at", endedAtIso) else put("p_ended_at", JsonNull)
            if (perceivedIntensity != null) {
                put("p_perceived_intensity", perceivedIntensity)
            } else {
                put("p_perceived_intensity", JsonNull)
            }
        }
        val res = supabase.postgrest.rpc(BackendContracts.Rpc.UPDATE_STRENGTH_WORKOUT_V1, params) { }
        return json.decodeFromString(res.data)
    }

    suspend fun finishStrengthWorkoutV1(
        supabase: SupabaseClient,
        workoutId: Int,
        endedAtIso: String,
        pausedSec: Int,
        hostExercises: List<StrengthFinishExercisePayload>,
        linked: List<StrengthFinishLinkedPayload> = emptyList()
    ): List<StrengthWorkoutSaveResultWire> {
        val params = buildJsonObject {
            put("p_workout_id", workoutId)
            put("p_ended_at", endedAtIso)
            put("p_paused_sec", pausedSec)
            put("p_exercises", StrengthWorkoutFinishCollapse.exercisesToJsonArray(hostExercises))
            put(
                "p_linked",
                buildJsonArray {
                    linked.forEach { link ->
                        add(
                            buildJsonObject {
                                put("workout_id", link.workoutId)
                                put("exercises", StrengthWorkoutFinishCollapse.exercisesToJsonArray(link.exercises))
                            }
                        )
                    }
                }
            )
        }
        val res = supabase.postgrest.rpc(BackendContracts.Rpc.FINISH_STRENGTH_WORKOUT_V1, params) { }
        return decodeFlexibleList(res.data)
    }

    private inline fun <reified T> decodeFlexibleList(raw: String): List<T> {
        val root = json.parseToJsonElement(raw)
        return when (root) {
            is JsonArray -> root.map { json.decodeFromString<T>(it.toString()) }
            is JsonObject -> listOf(json.decodeFromString<T>(root.toString()))
            else -> emptyList()
        }
    }
}

internal data class StrengthFinishLinkedPayload(
    val workoutId: Int,
    val exercises: List<StrengthFinishExercisePayload>
)
